### PR TITLE
feat: claimed total rewards

### PR DESCRIPTION
**Pull request by wedancedalot**
+ в оцей індекс додам поле claimed_total_rewards
+ add claimed_total_amount to pool
+ додати якусь перевірку що після 2 рівня на апгрейд най падає помилка
+ додати max stakers

TODO: add migration

### CODE REVIEW

File: programs/rewards/src/state/mining.rs [hunk 1]
```diff
@@ -177,9 +178,11 @@ pub struct RewardIndex {
     pub reward_mint: Pubkey,
     /// Rewards amount
     pub rewards: u64,
+    /// claimed_total_rewards
+    pub claimed_total_rewards: u64,
```
Comment (Messer4) [inline programs/rewards/src/state/mining.rs:182]: Just to think about

Could not be critical, but 
For high precision tokens, or tokens with extremely high emission u64 size could be not enough

File: programs/rewards/src/state/mining.rs [hunk 2]
```diff
@@ -74,6 +74,7 @@ impl Mining {
         let reward_index = self.reward_index_mut(reward_mint);
         let amount = reward_index.rewards;
         reward_index.rewards = 0;
+        reward_index.claimed_total_rewards += amount;
```
Comment (Messer4) [inline programs/rewards/src/state/mining.rs:77]: Suggest to use method with check for overflow

File: programs/rewards/src/state/reward_pool.rs [hunk 3]
```diff
@@ -55,15 +76,33 @@ impl RewardPool {
...
+        &mut self,
+        mining: &mut Mining,
+        amount: u64,
+        is_first_deposit: bool,
+        timestamp: u64,
+    ) -> ProgramResult {
         mining.refresh_rewards(self.vaults.iter(), timestamp)?;
 
+        if is_first_deposit {
+            if self.max_stakers > 0 && self.total_stakers >= self.max_stakers {
```
[the remark this answers is not in the document]
Comment (wedancedalot) [reply programs/rewards/src/state/reward_pool.rs:99]: this is to limit stakers ONLY if max_stakers is set

File: programs/rewards/src/state/mining.rs (hunk 1 above)
Comment (wedancedalot) [reply programs/rewards/src/state/mining.rs:182]: i don't think this is the case, this is solana TokenAccount struct and has u64
https://github.com/solana-labs/solana-program-library/blob/master/token/program/src/state.rs#L95

**Pull request merged**